### PR TITLE
Performance Improvement to SCIM Role PATCH operation

### PR DESCRIPTION
**Pull request by GANGANI**
## Purpose
Resolves: https://github.com/wso2/product-is/issues/15154
Related PRs: https://github.com/wso2/carbon-identity-framework/pull/4342, https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/449

### CODE REVIEW

File: modules/charon-core/src/main/java/org/wso2/charon3/core/protocol/endpoints/RoleResourceManager.java [hunk 1]
```diff
@@ -369,6 +378,11 @@ public SCIMResponse updateWithPATCHRole(String id, String patchRequest, RoleMana
             SCIMResourceTypeSchema schema = SCIMResourceSchemaManager.getInstance().getRoleResourceSchema();
             Map<String, Boolean> requestAttributes = ResourceManagerUtil.getAllAttributeURIs(schema);
 
+            List<PatchOperation> opList = getDecoder().decodeRequest(patchRequest);
```
Comment (AnuradhaSK) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/protocol/endpoints/RoleResourceManager.java:381]: Any chance of getting null for opList? If so handle the possible NPE

File: modules/charon-core/src/main/java/org/wso2/charon3/core/protocol/endpoints/RoleResourceManager.java [hunk 2]
```diff
@@ -526,4 +540,270 @@ public SCIMResponse updateWithPATCH(String existingId, String patchRequest, User
...
+    private boolean isUpdateAllUsersOperationFound(List<PatchOperation> patchOperations) throws JSONException {
+
+        for (PatchOperation patchOperation : patchOperations) {
+            String operation = patchOperation.getOperation();
+            String path = patchOperation.getPath();
+            JSONObject valuesJson = null;
+            if (StringUtils.isBlank(path)) {
+                valuesJson = (JSONObject) patchOperation.getValues();
+            }
+            if (operation.equals(SCIMConstants.OperationalConstants.REPLACE) &&
```
Comment (AnuradhaSK) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/protocol/endpoints/RoleResourceManager.java:560]: follow `<Constant>.equals(<variable>)` format
```suggestion
            if (SCIMConstants.OperationalConstants.REPLACE.equals(operation) &&
```

File: modules/charon-core/src/main/java/org/wso2/charon3/core/protocol/endpoints/RoleResourceManager.java [hunk 3]
```diff
@@ -526,4 +540,270 @@ public SCIMResponse updateWithPATCH(String existingId, String patchRequest, User
...
+                    (SCIMConstants.RoleSchemaConstants.USERS.equals(path) ||
+                            (valuesJson != null && valuesJson.has(SCIMConstants.RoleSchemaConstants.USERS)))) {
+                return true;
+            } else if (operation.equals(SCIMConstants.OperationalConstants.REMOVE) &&
```
Comment (AnuradhaSK) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/protocol/endpoints/RoleResourceManager.java:564]:
```suggestion
            } else if (SCIMConstants.OperationalConstants.REMOVE.equals(operation) &&
```

Comment (GANGANI) [reply modules/charon-core/src/main/java/org/wso2/charon3/core/protocol/endpoints/RoleResourceManager.java]: Fixed

File: modules/charon-core/src/main/java/org/wso2/charon3/core/protocol/endpoints/RoleResourceManager.java (hunk 2 above)
Comment (GANGANI) [reply modules/charon-core/src/main/java/org/wso2/charon3/core/protocol/endpoints/RoleResourceManager.java]: Fixed

File: modules/charon-core/src/main/java/org/wso2/charon3/core/protocol/endpoints/RoleResourceManager.java (hunk 1 above)
Comment (GANGANI) [reply modules/charon-core/src/main/java/org/wso2/charon3/core/protocol/endpoints/RoleResourceManager.java:381]: Fixed

**Review by AnuradhaSK — APPROVED** [review]

**Review by AnuradhaSK — APPROVED** [review]

**Pull request merged**